File: client_web/js/websocket_impl.js
const IsoWebSocket = require('isomorphic-ws');

class WebSocketClient {
  start(host, port, onConnected) {
    this.socket = new IsoWebSocket(`${host}:${port}`);
    var socket = this.socket;
    this.socket.onopen = function () {
      onConnected(new WebSocket(socket));
    };
  }

  close() {
    this.socket.close();
  }
}

class WebSocket {
  constructor(socket) {
    this.socket = socket;
  }

  send(data) {
    this.socket.send(data);
  }

  listen(onMessage) {
    this.socket.onmessage = function (event) {
      var reader = new FileReader();
      reader.onload = () => {
        var arrayBuffer = reader.result;
        var bytes = new Uint8Array(arrayBuffer);
        onMessage(bytes);
      };
      reader.readAsArrayBuffer(event.data);
    };
  }

  close() {
    this.socket.close();
  }
}

module.exports = {
  WebSocketClient,
  WebSocket
};

File: client_web/dart/lib/comms/websocket/websocket_impl.dart
@JS()
library websocket_js_impl;

import 'package:multiplayer_asteroids_common/common.dart';
import 'package:js/js.dart';

@JS()
class WebSocketClient implements CommsClient {
  external factory WebSocketClient();
  external void start(
    String host,
    int port,
    void onConnected(WebSocket socket),
  );
  external void close();
}

@JS()
class WebSocket implements Socket {
  external factory WebSocket();
  external void send(data);
  external void listen(void onMessage(data));
  external void close();
}
